### PR TITLE
WebGL1 fallback if WebGL2 context creation fails

### LINKED ISSUE #11819: WebGL version detection is not robust
Emscripten uses the following method to check whether WebGL2 is available:

https://github.com/emscripten-core/emscripten/blob/eb0ede77ae02e961cf8a266d55bdb513b6f03c23/src/library_browser.js#L321

However, sometimes `WebGL2RenderingContext` is available but WebGL2 is disabled due to hardware-specific reasons. For example, Moto E5, which runs the latest Chrome (84 as of this writing) and has the `WebGL2RenderingContext` type, disables WebGL2 dynamically. This leads Emscripten to believe that WebGL2 is available, ultimately leading it to this line:

https://github.com/emscripten-core/emscripten/blob/eb0ede77ae02e961cf8a266d55bdb513b6f03c23/src/library_browser.js#L347

A more robust approach would be to attempt to get a WebGL2 context, and if that fails (by returning `null`), attempt to get a WebGL1 context, and only if that fails too return null.

I discovered this while debugging a Flutter app that runs on top of CanvasKit. It's not a blocking issue as we can implement our own WebGL version detection and pass `majorVersion` to Emscripten.

### PR DESCRIPTION
**Pull request by kripken**
If WebGL2RenderingContext exists we still may fail to actually
create a context. To handle that, if both 1 and 2 are allowed in the
build, and 2 fails, try again with 1.

This increases code size a little, but as it is a correctness fix I
think it's unavoidable. Unless there is a better way?

helps #11819

### CODE REVIEW

**Comment by juj** [comment]
This has two problems: first is the increase in code size, like noticed. But second is that after this change, applications that would initialize a WebGL 2 context using `majorVersion = 2;`, may still get a successful WebGL context initialized, and be surprised that they are actually getting a WebGL 1 context, which the application might not work properly against.

Previously, the intended form has been that applications that are both WebGL 1 & 2 capable should attempt to first initialize a WebGL 2 context, and if that fails, then re-try initialization against WebGL 1. This aligns with native EGL/WGL/GLX/CoreGL(AppKit) APIs that do not either automatically give a lower tier graphics API than what was requested.

I.e. the ideal form to initialize a WebGL 1 or 2 context for applications that are able to simultaneously target both looks like

```c++
EMSCRIPTEN_WEBGL_CONTEXT_HANDLE CreateGLContext()
{
  EmscriptenWebGLContextAttributes attrs;
  emscripten_webgl_init_context_attributes(&attrs);
  for(int version = 2; version >= 1; --version)
  {
    attrs.majorVersion = version;
    EMSCRIPTEN_WEBGL_CONTEXT_HANDLE context = emscripten_webgl_create_context("#canvas", &attrs);
    if (context) return context;
  }
  return 0;
}
```
That way code size is not increased for users who do not co-target both APIs, and the caller will get only the context version that one was explicitly asking.

**Comment by kripken** [comment]
Thanks @juj, makes sense. I rewrote this PR to document that behavior, and to test it works.

**Comment by kripken** [comment]
I'm not sure. If the user didn't specify the version, and built with support for both 1 or 2, I can see an argument for at least trying 2, if 2 isn't 100% going to fail, which is what the current code does. But I see your side too, there is something unpredictable about doing it that way.

This PR just keeps the status quo and tries to document and test it. I don't have a strong preference for the current behavior, but @juj did?

File: src/settings.js [hunk 1]
```diff
@@ -530,6 +530,11 @@ var MIN_WEBGL_VERSION = 1;
 // Specifies the highest WebGL version to target. Pass -s MAX_WEBGL_VERSION=2
 // to enable targeting WebGL 2. If WebGL 2 is enabled, some APIs (EGL, GLUT, SDL)
 // will default to creating a WebGL 2 context if no version is specified.
+// Note that there is no automatic fallback to WebGL1 if WebGL2 is not supported
+// in the browser at runtime, even if you build with both WebGL1 and WebGL2
```
Comment (kainino0x) [inline src/settings.js:534]: "supported in the browser at runtime" is slightly unclear; how about:
```suggestion
// by the user's device, even if you build with both WebGL1 and WebGL2
```

**Comment by kripken** [comment]
Ah right, I'll remove the "fixes".

**Comment by juj** [comment]
> This doesn't do anything about the original issue, which was that `typeof WebGL2RenderingContext !== 'undefined'` is not a valid way to test for WebGL 2 support. If we're going to commit to requiring apps to do their own version fallback, shouldn't we remove that logic?
> 
> ```diff
> -          majorVersion: (typeof WebGL2RenderingContext !== 'undefined') ? 2 : 1,
> +          majorVersion: 2,
> ```



> I'm not sure. If the user didn't specify the version, and built with support for both 1 or 2, I can see an argument for at least trying 2, if 2 isn't 100% going to fail, which is what the current code does. But I see your side too, there is something unpredictable about doing it that way.
> 
> This PR just keeps the status quo and tries to document and test it. I don't have a strong preference for the current behavior, but @juj did?

Sorry, missed this conversation.

The function `Browser.createContext()` is something to consider internal for legacy/deprecated/emulation in general for other legacy APIs (Emscripten's JS SDL1, GLEW, GLUT, X11, EGL) that are supported. Some of those APIs do not have (implemented) a direct way of specifying which WebGL version to target, or the emulated APIs do have that kind of "pick highest available" as default. No user code should be using `Browser.createContext()` directly - but they are all using the HTML5 WebGL context creation API, or they are using one of SDL1, GLEW, GLUT, X11, EGL. If using one of those, they may be depending on the behavior that specifying WebGL version on the command line will also choose the version to target.

If someone wants to, it is fine to improve the support check into `Browser.createContext()` to actually do a feature check - but that functionality should be strictly done in the function `Browser.createContext()` and not in the "core" HTML5 WebGL creation API, since the core context creation API is expected to only initialize the version that was asked, or fail.

Though I presume improving that support is probably diminishing purpose work, the better approach would be to work towards the direction deleting `Browser` object and library_browser.js altogether.

**Pull request merged**